website/content/en/docs/advanced-topics/custom-bundle-validation.md

```diff
@@ -0,0 +1,315 @@
+}
+
+// FindBundleMetadata walks bundleRoot searching for metadata (ex. annotations.yaml),
+// and returns metadata and its path if found. If one is not found, an error is returned.
+func FindBundleMetadata(bundleRoot string) (Labels, string, error) {
+	return findBundleMetadata(afero.NewOsFs(), bundleRoot)
+}
+
+func findBundleMetadata(fs afero.Fs, bundleRoot string) (Labels, string, error) {
+	// Check the default path first, and return annotations if they were found or an error if that error
+	// is not because the path does not exist (it exists or there was an unmarshalling error).
+	annotationsPath := filepath.Join(bundleRoot, registrybundle.MetadataDir, registrybundle.AnnotationsFile)
+	annotations, err := readAnnotations(fs, annotationsPath)
+	if (err == nil && len(annotations) != 0) || (err != nil && !errors.Is(err, os.ErrNotExist)) {
+		return annotations, annotationsPath, err
+	}
+
+	// Annotations are not at the default path, so search recursively.
+	annotations = make(Labels)
+	annotationsPath = ""
+	err = afero.Walk(fs, bundleRoot, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		// Skip directories and hidden files, or if annotations were already found.
+		if len(annotations) != 0 || info.IsDir() || strings.HasPrefix(path, ".") {
+			return nil
+		}
+
+		annotationsPath = path
+		// Ignore this error, since we only care if any annotations are returned.
+		if annotations, err = readAnnotations(fs, path); err != nil {
+			log.Debug(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, "", err
+	}
+
+	if len(annotations) == 0 {
+		return nil, "", MetadataNotFoundError(bundleRoot)
+	}
+
+	return annotations, annotationsPath, nil
+}
+
+// readAnnotations reads annotations from file(s) in bundleRoot and returns them as Labels.
+func readAnnotations(fs afero.Fs, annotationsPath string) (Labels, error) {
+	// The annotations file is well-defined.
+	b, err := afero.ReadFile(fs, annotationsPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use the arbitrarily-labelled bundle representation of the annotations file
+	// for forwards and backwards compatibility.
+	annotations := registrybundle.AnnotationMetadata{
+		Annotations: make(Labels),
+	}
+	if err = yaml.Unmarshal(b, &annotations); err != nil {
+		return nil, fmt.Errorf("error unmarshalling potential bundle metadata %s: %v", annotationsPath, err)
+	}
+
+	return annotations.Annotations, nil
+}
+```
+
+The `main.go` is then built into a binary and used with `operator-sdk bundle
+validate`
+
+```sh
+$ go build -o myvalidator/main myvalidator/main.go
+$ operator-sdk bundle validate ./bundle --alpha-select-external ./myvalidator/main
+```
+```
+WARN[0000] Warning: Value sandbox-op.v0.0.1: owned CRD "sandboxes.sandbox.example.come" has an empty description
+INFO[0000] All validation tests have completed successfully
+```
+[errors-pkg]: https://github.com/operator-framework/api/pkg/validation/errors
```

Comment:
Link seems to be broken.

Reply:
```suggestion
[errors-pkg]: https://github.com/operator-framework/api/tree/master/pkg/validation/errors
```

Reply:
Yes, not able to access the errors-pkg link. Rest Looks good to me.

Reply:
I blindly used the pkg import from go rather than the actual human link :/ Fixed!